zookeeper-server/src/main/java/org/apache/zookeeper/common/FileChangeWatcher.java

```diff
@@ -0,0 +1,191 @@
+    private State state; // protected by synchronized(this)
+
+    /**
+     * Creates a watcher that watches <code>dirPath</code> and invokes <code>callback</code> on changes.
+     *
+     * @param dirPath the directory to watch.
+     * @param callback the callback to invoke with events. <code>event.kind()</code> will return the type of event,
+     *                 and <code>event.context()</code> will return the filename relative to <code>dirPath</code>.
+     * @throws IOException if there is an error creating the WatchService.
+     */
+    public FileChangeWatcher(Path dirPath, Consumer<WatchEvent<?>> callback) throws IOException {
+        FileSystem fs = dirPath.getFileSystem();
+        WatchService watchService = fs.newWatchService();
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Registering with watch service: " + dirPath);
+        }
+        dirPath.register(
+                watchService,
+                new WatchEvent.Kind<?>[]{
+                        StandardWatchEventKinds.ENTRY_CREATE,
+                        StandardWatchEventKinds.ENTRY_DELETE,
+                        StandardWatchEventKinds.ENTRY_MODIFY,
+                        StandardWatchEventKinds.OVERFLOW},
+                SensitivityWatchEventModifier.HIGH);
+        state = State.NEW;
+        this.watcherThread = new WatcherThread(watchService, callback);
+        this.watcherThread.setDaemon(true);
+    }
+
+    public synchronized State getState() {
+        return state;
+    }
+
+    private synchronized void setState(State newState) {
+        state = newState;
+        this.notifyAll();
+    }
+
+    /**
+     * Tells the background thread to start. Does not wait for it to be running.
+     * Calling this method more than once has no effect.
+     */
+    public synchronized void start() {
+        if (state != State.NEW) {
+            return;
+        }
+        setState(State.STARTING);
+        this.watcherThread.start();
+    }
+
+    /**
+     * Tells the background thread to stop. Does not wait for it to exit.
+     */
+    public synchronized void stop() {
+        if (state == State.STARTING || state == State.RUNNING) {
+            setState(State.STOPPING);
+            watcherThread.interrupt();
+        }
+    }
+
+    /**
+     * Inner class that implements the watcher thread logic.
+     */
+    private class WatcherThread extends ZooKeeperThread {
+        private static final String THREAD_NAME = "FileChangeWatcher";
+
+        final WatchService watchService;
+        final Consumer<WatchEvent<?>> callback;
+
+        WatcherThread(WatchService watchService, Consumer<WatchEvent<?>> callback) {
+            super(THREAD_NAME);
+            this.watchService = watchService;
+            this.callback = callback;
+        }
+
+        @Override
+        public void run() {
+            try {
+                LOG.info(getName() + " thread started");
+                synchronized (FileChangeWatcher.this) {
```

Comment:
We have to check the state and modify it within one critical section. If I change the code to this:

```
if (FileChangeWatcher.this.getState() == FileChangeWatcher.State.STARTING) {
  FileChangeWatcher.this.setState(FileChangeWatcher.State.RUNNING);
}
```

then there is a race between me checking the state and changing it. During this time, another thread could call `stop()` and change the state to `STOPPING`. Then I would overwrite `STOPPING` with `RUNNING` and suppress the `stop()` command.

I should probably document this in a comment so it doesn't get refactored into incorrect code later by someone else.

Reply:
My intention was, checking and state transition logic can be moved to setState as It is the entry point to change, state transition needs to be managed here.

The below can be moved to setState

if (FileChangeWatcher.this.getState() == FileChangeWatcher.State.STARTING) {
 state = FileChangeWatcher.State.RUNNING;
}

we can simply check state is RUNNING or not to proceed further in the run method.

Reply:
I don't think that logic belongs in `setState()` since it's just a simple setter method. Are you proposing I add a new method that's something like "compareAndSetState(State expected, State newState)"?

Reply:
Yes.Kind of state matrix->(current state,new state)->derived state

Reply:
Please see the latest version.

Reply:
I see it. Thanks for the change.